src/client/Client.cc

```diff
@@ -6368,7 +6368,7 @@ void Client::handle_command_reply(const MConstRef<MCommandReply>& m)
 
   auto &op = command_table.get_command(tid);
   if (op.outbl) {
-    *op.outbl = m->get_data();
```

Comment:
https://tracker.ceph.com/issues/61819#note-4

Good catch. Another question: What if one of the ranks return an error? How is that handled?

Reply:
Venky Shankar wrote:
> Jinmyeong Lee wrote:
> > Jinmyeong Lee wrote:
> > > https://github.com/ceph/ceph/pull/52200
> > 
> > I also want to discuss how to append the result in bufferlist.
> > From now, I just append the result in the bufferlist, but I think the below format would be better. (Also afraid of side-effect)
> > 
> > {
> >     tid_1: [],
> >     tid_2: [],
> >     ...
> > }
> 
> Is the `tid_*` prefix the transaction-id? IMO, using mds rank or IDs would be better.

I agree. To make a similar result with a result of "ceph tell mds.* session ls" command.

// For all active MDSs:
{
    "mds.$id_0": [],
    "mds.$id_1": [],
...
}

I think if some ranks return error, then it is okay to skip to add to the outbl. -> I will add a commit for this.
When I send "mds.* session ls" command using ceph CLI, the standby MDSs return -22(EINVAL) and the result is not in CLI command.

But to avoid a side effects about a single target mds command, I think it would be good to add a new multi_target_id to common/CommandTable to confirm whether all the commands end. (so it can make possible to make the above format of result.)

So,
In common/CommandTable,
* Add new functions: ceph_tid_t get_new_multi_command_id(), bool is_all_multi_commands_end(ceph_tid_t multi_target_id)
* Adjust: T& start_command(ceph_tid_t multi_target_id=0)
* Add a few lines of code in Client::mds_command and Client::handle_command_reply
- mds_command
```
ceph_tid_t multi_target_id = 0;
  if (non_laggy.size() > 1) {
      std::scoped_lock cmd_lock(command_lock);
      multi_target_id = command_table.get_new_multi_command_id();
  }
...
```
- handle_command_reply: append the command result to bufferlist with format

Reply:
@vshankar PING: Could you give feedback about this approach?

Reply:
Sure. Sorry - was on PTO. Just back. Will have a look.

Reply:
> Venky Shankar wrote:
> 
> > Jinmyeong Lee wrote:
> > > Jinmyeong Lee wrote:
> > > > #52200
> > > 
> > > 
> > > I also want to discuss how to append the result in bufferlist.
> > > From now, I just append the result in the bufferlist, but I think the below format would be better. (Also afraid of side-effect)
> > > {
> > > tid_1: [],
> > > tid_2: [],
> > > ...
> > > }
> > 
> > 
> > Is the `tid_*` prefix the transaction-id? IMO, using mds rank or IDs would be better.
> 
> I agree. To make a similar result with a result of "ceph tell mds.* session ls" ceph CLI command.
> 
> // For all active MDSs: { "mds.$id_0": [], "mds.$id_1": [], ... }

Looks good.

> 
> I think if some ranks return error, then it is okay to skip to add to the outbl. -> I will add a commit for this. When I send "mds.* session ls" command using ceph CLI, the standby MDSs return -22(EINVAL) and the result is not in CLI command.

Should we add the failed errno as part of the json? Which means for each mds reply the json will have an errno and the command output. If errno is not zero, then the command output would be empty. What do you think?

> 
> But to avoid side effects about a single target mds command, I think it would be good to add a new multi_target_id to common/CommandTable to confirm whether all the commands end. (so it can make it possible to make the above format of the result.)

You mean the single target mds command now using the changed format, right?

Reply:
> You mean the single target mds command now using the changed format, right?
I mean, No.
If we edit the result format of a single MDS target command, we should consider other parts using this Client like here(https://github.com/ceph/ceph/blob/main/src/ceph.in#L1252C20-L1254). I am a little afraid of these side effects.
Also, the result of "ceph tell mds.0 session ls" Ceph CLI command is only a list of sessions. (There is no key with "mds.0".) I think following the Ceph CLI results would be better.
So, I want to make a new format when just the mds_spec is "*", which means multi-target MDSs.


> Should we add the failed errno as part of the json? Which means for each mds reply the json will have an errno and the command output. If errno is not zero, then the command output would be empty. What do you think?

I briefly considered if it is possible, but in my opinion, it looks difficult.
In libcephfs, when "ceph_mds_command" function calls the "Client::mds_command" with mds_spec = "*", all command Messages are on one C_Safer_Cond.

And in "Client::handle_command_reply" function, the individual response from each MDS is handled and the errno is updated in "complete(r)". So in this situation, the libcephfs can get only one errorno from "cond.wait()".
(
You can check it here:
https://github.com/ceph/ceph/blob/main/src/libcephfs.cc#L543-L555
https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L6334-L6347
https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L6389-L6391
)

If we should keep all the errornos, we should introduce a new field to keep each errono in somewhere, like in common/CommandTable. Or introducing somewhat new Structure for multi-target commands? 😀

Reply:
> Thanks for your feedback!!
> 
> > You mean the single target mds command now using the changed format, right?
> 
> I mean, No. If we edit the result format of a single MDS target command, we should consider other parts using this Client like here(https://github.com/ceph/ceph/blob/main/src/ceph.in#L1252C20-L1254). I am a little afraid of these side effects. Also, the result of "ceph tell mds.0 session ls" Ceph CLI command is only a list of sessions. (There is no key with "mds.0".) I think following the Ceph CLI results would be better. So, I want to make a new format when just the mds_spec is "*", which means multi-target MDSs.

Right - makes sense. I'm fine with your proposal of adding multi_target_it.

> 
> > Should we add the failed errno as part of the json? Which means for each mds reply the json will have an errno and the command output. If errno is not zero, then the command output would be empty. What do you think?
> 
> I briefly considered if it is possible, but in my opinion, it looks difficult. In libcephfs, when "ceph_mds_command" function calls the "Client::mds_command" with mds_spec = "*", all command Messages are on one C_Safer_Cond.

Ah, that's a bummer.

> 
> And in "Client::handle_command_reply" function, the individual response from each MDS is handled and the errno is updated in "complete(r)". So in this situation, the libcephfs can get only one errorno from "r = cond.wait()". ( You can check it here: https://github.com/ceph/ceph/blob/main/src/libcephfs.cc#L543-L555 https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L6334-L6347 https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L6389-L6391 )
> 
> If we should keep all the errornos, we should introduce a new field to keep each errono in somewhere, like in 
common/CommandTable. Or introducing somewhat new Structure for multi-target commands? 😀

Anyway, let's stick to your original idea.

Reply:
Thanks for checking! I will make commits for this ASAP.